pkg/resource/training_job/custom_set_update_input.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Use this file if conditions need to be updated based on the latest status
+// of training job which is not evident from API response
+
+package training_job
+
+import (
+	"errors"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+	svcapitypes "github.com/aws-controllers-k8s/sagemaker-controller/apis/v1alpha1"
+	svcsdk "github.com/aws/aws-sdk-go/service/sagemaker"
+)
+
+// buildProfilerRuleConfigUpdateInput sets the input of the ProfilerRuleConfiguration so that
+// it is compatible with the sagemaker API.
+// Update training job is post operation wrt to the profiler parameters.
+// Because of this only NEW rules can be specified.
+// In this function we check to see if any new profiler configurstions have been added.
+func buildProfilerRuleConfigUpdateInput(desired *resource, latest *resource, input *svcsdk.UpdateTrainingJobInput) error {
+	profilerRuleDesired := desired.ko.Spec.ProfilerRuleConfigurations
+	profilerRuleLatest := latest.ko.Spec.ProfilerRuleConfigurations
+
+	if ackcompare.IsNil(profilerRuleLatest) {
+		return nil
+	}
+	if len(profilerRuleDesired) < len(profilerRuleLatest) {
```

Comment:
Dumb question, but if its possible that len(profilerRuleDesired) might be less than latest, can profilerRuleDesired be nil and this method still be called somehow? I'm assuming no

Reply:
No, if profilerRuleDesired is nil the profiler is removed. https://github.com/aws-controllers-k8s/sagemaker-controller/blob/d7f828a94c30f11f43dac1ce3ab472f657c68d6c/pkg/resource/training_job/sdk.go#L1165

Reply:
this is pretty basic check and is not robust, if a customer changes the rules instead of adding new ones or removed few of them, current logic will fail after the update, it will add the new rules to the job but still find a diff afterwards.

@ananth102 you already mentioned it in the comments but maybe worth clarifying the expectations further for future reference

I intentionally ignored it for now to keep the focus on adding rules when the customer wants to instead of covering all possibilities based on service behavior. We can do one simple thing to make it a bit more intelligent instead of comparing the length, mark the rules with 0 when you find from latest and mark them 1 when you find it in desired. After the loop, if all the ruleMap entries are not 1, error saying existing rules cannot be removed. wdyt?

also change the file name to custom_update if you plan on keeping it

Reply:
Sounds like there are four cases:
1. Customer adds rule (Happy)
2. Customer removes rule (terminal)
3. Customer removes some rules and adds other -> Controller would try/fail updates and eventually be set to terminal
4. Customer changes rule -> Controller would try/fail updates and eventually be set to terminal

That approach in your comment solves the first three cases. For the fourth one I can change the < in line 40 to <= and return a regular error instead of a terminal error. It will eventually terminate once the training job completes.